python/cudf/cudf/core/dataframe.py

```diff
@@ -2897,96 +2897,201 @@ def add_column(self, name, data, forceindex=False):
+               weight  250.0  150.0
+        falcon speed   320.0  250.0
+               weight    1.0    0.8
         """
-        if axis == 0 and labels is not None:
-            raise NotImplementedError("Can only drop columns, not rows")
-        if errors != "raise":
-            raise NotImplementedError("errors= keyword not implemented")
-        if labels is None and columns is None:
-            raise ValueError(
-                "Need to specify at least one of 'labels' or 'columns'"
-            )
-        if labels is not None and columns is not None:
-            raise ValueError("Cannot specify both 'labels' and 'columns'")
 
         if labels is not None:
+            if index is not None or columns is not None:
+                raise ValueError(
+                    "Cannot specify both 'labels' and 'index'/'columns'"
+                )
             target = labels
-        else:
+        elif index is not None:
+            target = index
+            axis = 0
+        elif columns is not None:
             target = columns
+            axis = 1
+        else:
+            raise ValueError(
+                "Need to specify at least one of 'labels', "
+                "'index' or 'columns'"
+            )
 
-        columns = (
-            [target]
-            if isinstance(target, (str, numbers.Number))
-            else list(set(target))
-        )
         if inplace:
             outdf = self
         else:
             outdf = self.copy()
-        for c in columns:
-            outdf._drop_column(c)
-        return outdf
 
-    def drop_column(self, name):
-        """Drop a column by *name*
-        """
-        warnings.warn(
-            "The drop_column method is deprecated. "
-            "Use the drop method instead.",
-            DeprecationWarning,
-        )
-        self._drop_column(name)
+        if axis in (1, "columns"):
+            target = _get_host_unique(target)
+
+            _drop_columns(outdf, target, errors)
+        elif axis in (0, "index"):
+            if not isinstance(target, (cudf.Series, cudf.Index)):
+                target = column.as_column(target)
+
+            if isinstance(self._index, cudf.MultiIndex):
+                if level is None:
+                    level = 0
+
+                levels_index = outdf.index.get_level_values(level)
+                if errors == "raise" and not target.isin(levels_index).all():
+                    raise KeyError("One or more values not found in axis")
+
+                # TODO : Could use anti-join as a future optimization
+                sliced_df = outdf.take(~levels_index.isin(target))
+                sliced_df._index.names = self._index.names
+            else:
+                if errors == "raise" and not target.isin(outdf.index).all():
+                    raise KeyError("One or more values not found in axis")
+
+                sliced_df = outdf.join(
+                    cudf.DataFrame(index=target), how="leftanti"
+                )
```

Comment:
I know I suggested this, but have we confirmed that this maintains the order as expected when dropping rows?

Reply:
Yes

```python
>>> import cudf
>>> import numpy as np
>>> df = cudf.DataFrame({'A':[1, 1, 2, 2] * 100,
...   'B': [1, 2, 3, 4]*100,
...  'C': np.random.randn(400)})
>>> df
     A  B         C
0    1  1 -0.800933
1    1  2  0.231304
2    2  3  0.185854
3    2  4  0.382024
4    1  1 -1.321089
..  .. ..       ...
395  2  4  0.242435
396  1  1  1.245609
397  1  2  0.255610
398  2  3  1.085539
399  2  4 -1.277602

[400 rows x 3 columns]
>>> df.set_index(['A', 'B'])
            C
A B          
1 1 -0.800933
  2  0.231304
2 3  0.185854
  4  0.382024
1 1 -1.321089
..        ...
2 4  0.242435
1 1  1.245609
  2  0.255610
2 3  1.085539
  4 -1.277602

[400 rows x 1 columns]

>>> gx = df.set_index(['A', 'B'])
>>> px = gx.to_pandas()
>>> from cudf.tests.utils import assert_eq
/nvme/0/pgali/envs/cudfdev/lib/python3.7/site-packages/pandas/util/__init__.py:12: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing
>>> px.drop(index=1, level=0)
            C
A B          
2 3  0.185854
  4  0.382024
  3 -0.712118
  4 -0.089184
  3  0.262499
..        ...
  4 -1.705039
  3 -0.557653
  4  0.242435
  3  1.085539
  4 -1.277602

[200 rows x 1 columns]
>>> gx.drop(index=1, level=0)
            C
A B          
2 3  0.185854
  4  0.382024
  3 -0.712118
  4 -0.089184
  3  0.262499
..        ...
  4 -1.705039
  3 -0.557653
  4  0.242435
  3  1.085539
  4 -1.277602

[200 rows x 1 columns]
>>> assert_eq(gx.drop(index=1, level=0), px.drop(index=1, level=0))
True
>>> assert_eq(gx.drop(index=3, level=1), px.drop(index=3, level=1))
True
```